File: src/main/java/cspfj/generator/constraint/GeneratorManager.java
package cspfj.generator.constraint;

import java.util.HashMap;
import java.util.Map;

import cspfj.exception.FailedGenerationException;
import cspfj.problem.Problem;
import cspom.constraint.CSPOMConstraint;

public final class GeneratorManager {

    private static final Map<String, Class<? extends Generator>> KNOWN = new HashMap<String, Class<? extends Generator>>();

    static {
        register("abs", AbsGenerator.class);
        register("add", AddGenerator.class);
        register("sub", AddGenerator.class);
        register("alldifferent", AllDifferentGenerator.class);
        register("or", DisjGenerator.class);
        register("not", DisjGenerator.class);
        register("eq", EqGenerator.class);
        register("neg", EqGenerator.class);
        register("ext", ExtensionGenerator.class);
        register("gt", GtGenerator.class);
        register("ge", GtGenerator.class);
        register("lt", GtGenerator.class);
        register("le", GtGenerator.class);
        register("mul", MulGenerator.class);
        register("ne", NeqGenerator.class);
        register("absdiff", AbsDiffGenerator.class);
        register("diffGe", DiffGeGenerator.class);
    }

    private final Problem problem;
    private final Map<Class<? extends Generator>, Generator> generators;

    public GeneratorManager(final Problem problem) {
        generators = new HashMap<Class<? extends Generator>, Generator>();
        this.problem = problem;
    }

    public static void register(final String description,
            final Class<? extends Generator> clazz) {

        KNOWN.put(description, clazz);

    }

    public boolean generate(final CSPOMConstraint constraint)
            throws FailedGenerationException {
        final Class<? extends Generator> candidate = KNOWN.get(constraint
                .getDescription());

        if (candidate == null) {
            throw new FailedGenerationException("No candidate constraint for "
                    + constraint + " (" + constraint.getDescription() + ")");
        }

        Generator generator = generators.get(candidate);
        if (generator == null) {
            try {
                generator = candidate.getConstructor(Problem.class)
                        .newInstance(problem);
            } catch (Exception e) {
                throw new FailedGenerationException(e);
            }
            generators.put(candidate, generator);
        }

        return generator.generate(constraint);
    }

}

File: src/main/scala/cspfj/generator/constraint/GeneratorManager.scala
package cspfj.generator.constraint;

import cspfj.generator.FailedGenerationException
import cspfj.Problem
import cspom.constraint.CSPOMConstraint;
import scala.collection.mutable.HashMap

final class GeneratorManager(val problem: Problem) {

  var generators: HashMap[Class[_ <: AbstractGenerator], AbstractGenerator] = new HashMap();

  @throws(classOf[FailedGenerationException])
  def generate(constraint: CSPOMConstraint) = {
    val candidate = GeneratorManager.known.getOrElse(constraint.description.toLowerCase,
      throw new FailedGenerationException(s"No candidate constraint for $constraint"))

    generators.getOrElseUpdate(candidate, candidate.getConstructor(classOf[Problem]).newInstance(problem)).generate(constraint)
  }
}

object GeneratorManager {
  var known: Map[String, Class[_ <: AbstractGenerator]] = Map(
    "abs" -> classOf[AbsGenerator],
    "add" -> classOf[AddGenerator],
    "sub" -> classOf[AddGenerator],
    "alldifferent" -> classOf[AllDifferentGenerator],
    "or" -> classOf[DisjGenerator],
    "not" -> classOf[DisjGenerator],
    "and" -> classOf[DisjGenerator],
    "eq" -> classOf[EqGenerator],
    "neg" -> classOf[EqGenerator],
    "ext" -> classOf[ExtensionGenerator],
    "gt" -> classOf[GtGenerator],
    "ge" -> classOf[GtGenerator],
    "lt" -> classOf[GtGenerator],
    "le" -> classOf[GtGenerator],
    "mul" -> classOf[MulGenerator],
    "ne" -> classOf[NeqGenerator],
    "absdiff" -> classOf[AbsDiffGenerator],
    "diffge" -> classOf[DiffGeGenerator],
    "gcc" -> classOf[GccGenerator],
    "mod" -> classOf[ModGenerator],
    "nevec" -> classOf[NeqVecGenerator],
    "zerosum" -> classOf[ZeroSumGenerator],
    "lexleq" -> classOf[LexLeqGenerator])

  def register(entry: (String, Class[_ <: AbstractGenerator])) {
    known += entry
  }
}
